src/tr/uv/tr_uv_tcp_aux.c

```diff
@@ -788,62 +794,44 @@ void tcp__on_heartbeat(tr_uv_tcp_transport_t* tt)
     int rtt = 0;
     int start = 0;
 
-    if (!tt->is_waiting_hb) {
-        pc_lib_log(PC_LOG_WARN, "tcp__on_heartbeat - tcp is not waiting for heartbeat, ignore");
-        return;
-    }
-
-    pc_lib_log(PC_LOG_DEBUG, "tcp__on_heartbeat - received heartbeat from server");
+    pc_lib_log(PC_LOG_DEBUG, "tcp__on_heartbeat - [Heartbeat] received from server");
     pc_assert(tt->state == TR_UV_TCP_DONE);
-    pc_assert(uv_is_active((uv_handle_t*)&tt->hb_timeout_timer));
 
+    // FIXME, TODO(lhahn): consider removing rtt code, since it is probably broken
     /*
      * we hacking uv timer to get the heartbeat rtt, rtt in millisec
      * int is enough to hold the value
      */
-    start = (int)(tt->hb_timeout_timer.timeout - tt->hb_timeout * 1000);
-
     rtt = (int)(tt->uv_loop.time - start);
 
-    uv_timer_stop(&tt->hb_timeout_timer);
-
-    tt->is_waiting_hb = 0;
-
     if (tt->hb_rtt == -1) {
         tt->hb_rtt = rtt;
     } else {
         tt->hb_rtt = (tt->hb_rtt * 2 + rtt) / 3;
         pc_lib_log(PC_LOG_DEBUG, "tcp__on_heartbeat - calc rtt: %d ms", tt->hb_rtt);
+        
     }
-
-    uv_timer_start(&tt->hb_timer, tcp__heartbeat_timer_cb, tt->hb_interval * 1000, 0);
 }
 
 void tcp__heartbeat_timer_cb(uv_timer_t* t)
 {
     GET_TT(t);
-
     pc_assert(t == &tt->hb_timer);
-    pc_assert(tt->is_waiting_hb == 0);
     pc_assert(tt->state == TR_UV_TCP_DONE);
+    
+    uint64_t threshold = (tt->hb_interval * 1000) * (PC_HEARTBEAT_TIMEOUT_FACTOR + 4); // +1 here to keep with the old behaviour
+    // We check whether the server stopped sending packets to the client.
+    // If it did we trigger a reconnection and stop sending heartbeats.
+    uint64_t time_elapsed = uv_now(&tt->uv_loop) - tt->last_server_packet_time;
+    if (time_elapsed > threshold) {
+        pc_lib_log(PC_LOG_WARN, "tcp__heartbeat_timer_cb - heartbeat timeout, will reconn");
+        pc_trans_fire_event(tt->client, PC_EV_UNEXPECTED_DISCONNECT, "HB Timeout", NULL);
+        tt->reconn_fn(tt);
+        return;
+    }
 
     tcp__send_heartbeat(tt);
-    tt->is_waiting_hb = 1;
-    pc_lib_log(PC_LOG_DEBUG, "tcp__heartbeat_timer_cb - start heartbeat timeout timer");
-
-    uv_timer_start(&tt->hb_timeout_timer, tcp__heartbeat_timeout_cb, tt->hb_timeout * 1000, 0);
-}
-
-void tcp__heartbeat_timeout_cb(uv_timer_t* t)
-{
-    GET_TT(t);
-
-    pc_assert(tt->is_waiting_hb);
-    pc_assert(t == &tt->hb_timeout_timer);
-
-    pc_lib_log(PC_LOG_WARN, "tcp__heartbeat_timeout_cb - will reconn, hb timeout");
-    pc_trans_fire_event(tt->client, PC_EV_UNEXPECTED_DISCONNECT, "HB Timeout", NULL);
-    tt->reconn_fn(tt);
+    uv_timer_start(&tt->hb_timer, tcp__heartbeat_timer_cb, tt->hb_interval * 1000, 0);
```

Comment:
Why you are calling `uv_timer_start` every time? Why not using the `repeat` [option](http://docs.libuv.org/en/v1.x/timer.html#c.uv_timer_start)?